src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -392,5 +411,99 @@ public int GetArrayRank()
 #else
         ImmutableArray<TypeName> GetGenericArguments() => _genericArguments;
 #endif
+
+#if SYSTEM_REFLECTION_METADATA
+        /// <summary>
+        /// Creates a new <see cref="TypeName" /> object that represents current simple name with provided assembly name.
+        /// </summary>
+        /// <param name="assemblyName">Assembly name.</param>
+        /// <returns>Created simple name.</returns>
+        /// <exception cref="InvalidOperationException">The current type name is not simple.</exception>
+        public TypeName MakeSimpleTypeName(AssemblyNameInfo? assemblyName)
```

Comment:
> All other `Make` method take the `this` instance as a whole and make a new thing from it.

> This method is different. It replaces a property on `this` instance with new value, it does not take the `this` instance as a whole like all other methods.

It does that for a reason: I wanted to hide the complexity of changing nested type names from the user.

Let's condiser `CreateSimpleTypeName` so I can better express what I mean:

```cs
public static TypeName CreateSimpleTypeName(
        string metadataName,
        TypeName? declaringType = null,
        AssemblyNameInfo? assemblyInfo = null);
```

What if the input is following:
- `declaringType`: a type called `MyNamespace1.Declaring`
- `metadataName`: `MyNamespace2.Declaring+Nested`

We either end up with an invalid name (the declaring and nested type have to belong to the same namespace and assembly) or a need to verify them.

There is one more question: is the `metadataName` supposed to be a full name or just the name for nested types?

With the `public TypeName MakeSimpleTypeName(AssemblyNameInfo? assemblyInfo)` the users are allowed to change the `AssemblyName` and don't need to worry about the nested types (it's handled for them). There is no need to re-parse or validate anything and it is exactly what I need.

> The name does not look right to me.
> If we really need to have this method, `WithAssemblyName` that you have proposed previously sounds better to me.

Naming is hard and subjective. For me, the `MakeSimpleTypeName` name suggests that it works only for simple types and can be used to create a new instance with a different assembly name.

I also like the previous proposal (`WithAssemblyName`), but it does not suggest anything that it's going to work only for simple names.

> This method is a small optimization of the BF scenario. (I am not convinced that it is necessary to make NRBF secure.) 

The BF scenario is important to me and I am not convinced that I don't need it.

@jkotas If you believe that this name is wrong I am open to marking it as blocking and re-discussing it tomorrow with the API review board.

Reply:
> Do you agree that it is just a minor optimization, and that this exact API is not required to make the Nrbf-reader secure?

I don't agree. It's not a minor optimization, it is my requirement that was pointed out by @GrabYourPitchforks in code review (lack of any kind of re-parsing). Why otherwise would I be pushing for including it in RC1?

Reply:
Can you explain the rationale behind this requirement?

I think that the requirement should be to avoid any non-linear work (where N is the size of the input). For example, allocating a substring (that I expect that many use cases will end up doing) or re-validating whether the string escaped properly should be fine since it is still linear amount of work.